README.md
A maidenhead library for HAM radio operators. Converting lat/lon to maidenhead locator
and vice versa,...

[license](https://raw.githubusercontent.com/goranskular/maidenhead/master/LICENSE).

## Usage

A simple usage example:

```dart
import 'package:maidenhead/maidenhead.dart';

void main(List<String> arguments) {
  print(Maidenhead.to_location('JN75xu32SP')); // 45.844271  15.947917
  print(Maidenhead.to_location('JN75XU32NV')); // 45.845313 15.946181
  print(Maidenhead.to_maiden(45.844271, 15.947917, 5));
  print(Maidenhead.google_maps_maiden('JN75xu32SP'));

    // 177.69963804384477 meters
  print(Haversine.distance_maiden('JN75xu32SP', 'JN75XU32NV'));

  var pos1 = Maidenhead.latlon(45.844271, 15.947917, precision: 5);
  var pos2 = Maidenhead.latlon(45.844271, 15.947917, precision: 5);
  var distance = pos1.distance_to(pos2);
  var bearing = pos1.bearing_to(pos2);
  print('distance: ' + distance.toString() + ' bearing: ' + bearing.toString());
  print('pos1 to pos2 distance: ' +
      Haversine.distance_maiden(pos1.grid, pos2.grid).toString());
}
```

## Github

Please file feature requests and bugs at the [github](https://github.com/goranskular/maidenhead.git)

## pub.dev

[pub.dev](https://pub.dev/packages/maidenhead)

73 de 9A1GS (former 9A3GOS)

CHANGELOG.md
## 0.0.1-dev

- Initial version

## 0.0.2-dev

- Updated README.md

## 0.0.3-dev

- Comments added

## 0.0.4-dev

- Added bearing and Harversine distance calculator with examples

## 0.0.5-dev

- intl 0.16.1
- validate 1.7.0



test/maidenhead_test.dart
/* import 'package:maidenhead/maidenhead.dart';
import 'package:test/test.dart';

void main() {
  group('A group of tests', () {
    Awesome awesome;

    setUp(() {
      awesome = Awesome();
    });

    test('First Test', () {
      expect(awesome.isAwesome, isTrue);
    });
  });
}
*/


lib/maidenhead.dart
library maidenhead;

import 'dart:math' as math;
// for NumberFormat
import 'package:intl/intl.dart';

export 'src/maidenhead_base.dart';
export 'src/Haversine.dart';

/// Equator radius in meter (WGS84 ellipsoid)
const double EQUATOR_RADIUS = 6378137.0;

/// Polar radius in meter (WGS84 ellipsoid)
const double POLAR_RADIUS = 6356752.314245;

/// WGS84
const double FLATTENING = 1 / 298.257223563;

/// Earth radius in meter
const double EARTH_RADIUS = EQUATOR_RADIUS;

/// The PI constant.
const double PI = math.pi;

/// Rounds [value] to given number of [decimals]
double round(final double value, {final int decimals = 6}) =>
    (value * math.pow(10, decimals)).round() / math.pow(10, decimals);

/// Converts degree to radian
double degToRadian(final double deg) => deg * (PI / 180.0);

/// Radian to degree
double radianToDeg(final double rad) => rad * (180.0 / PI);

/// Convert a bearing to be within the 0 to +360 degrees range.
/// Compass bearing is in the rangen 0° ... 360°
double normalizeBearing(final double bearing) => (bearing + 360) % 360;

/// Converts a decimal coordinate value to sexagesimal format
///
///     final String sexa1 = decimal2sexagesimal(51.519475);
///     expect(sexa1, '51° 31\' 10.11"');
///
String decimal2sexagesimal(final double dec) {
  List<int> _split(final double value) {
    // NumberFormat is necessary to create digit after comma if the value
    // has no decimal point (only necessary for browser)
    //final List<String> tmp = new NumberFormat('0.0#####')
    var tmp =
        NumberFormat('0.0#####').format(round(value, decimals: 10)).split('.');
    return <int>[int.parse(tmp[0]).abs(), int.parse(tmp[1])];
  }

  var parts = _split(dec);
  var integerPart = parts[0];
  var fractionalPart = parts[1];

  var deg = integerPart;
  var min = double.parse('0.${fractionalPart}') * 60;

  var minParts = _split(min);
  var minFractionalPart = minParts[1];

  var sec = (double.parse('0.${minFractionalPart}') * 60);

  return "${deg}° ${min.floor()}' ${round(sec, decimals: 2).toStringAsFixed(2)}\"";
}


lib/src/Haversine.dart
import 'package:maidenhead/maidenhead.dart';
import 'dart:math' as math;

class Haversine {
  const Haversine();

  /// Calculates distance with Haversine algorithm.
  ///
  /// Accuracy can be out by 0.3%
  /// More on [Wikipedia](https://en.wikipedia.org/wiki/Haversine_formula)
  /// Original in https://github.com/MikeMitterer/dart-latlong/tree/master/lib/latlong/calculator
  static double distance(final double lat1, final double lon1,
      final double lat2, final double lon2) {
    final sinDLat = math.sin((degToRadian(lat2) - degToRadian(lat1)) / 2);
    final sinDLng = math.sin((degToRadian(lon2) - degToRadian(lon1)) / 2);

    // Sides
    final a = sinDLat * sinDLat +
        sinDLng *
            sinDLng *
            math.cos(degToRadian(lat1)) *
            math.cos(degToRadian(lat2));
    final c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a));

    return EQUATOR_RADIUS * c;
  }

  static double distance_maiden(String grid_p1, String grid_p2) {
    var p1 = Maidenhead.to_location(grid_p1);
    var p2 = Maidenhead.to_location(grid_p2);
    return distance(p1[0], p1[1], p2[0], p2[1]);
  }
}


lib/src/maidenhead_base.dart
import 'package:maidenhead/maidenhead.dart';
import 'dart:math' as math;

class Maidenhead {
  /// Latitude and Longitude stored in a double
  double lat, lon;

  /// Maidenead locator grid string
  String grid;

  /// Initialize from latlong and calculate maidenhead grid
  Maidenhead.latlon(double lat, double lon, {int precision = 3}) {
    this.lat = lat;
    this.lon = lon;
    grid = to_maiden(lat, lon, precision);
  }

  /// Initialize from grid and calculate lat and long
  Maidenhead.grid(String grid) {
    this.grid = grid;
    var latlon = to_location(grid);
    lat = latlon[0];
    lon = latlon[1];
  }

  /// Convert latitude and longitude to the maidenhead grid locator
  /// default most used precision is 3
  /// Precision 5 is very accurate
  static String to_maiden(double lat, double lon, [int precision = 3]) {
    //
    final nA = 'A'.codeUnits.first;
    var lon_div = ((lon + 180) / 20).floor().toDouble();
    var lon_remainder = ((lon + 180) % 20).toDouble();
    var lat_div = ((lat + 90) / 10).floor().toDouble();
    var lat_remainder = ((lat + 90) % 10).toDouble();
    // first pair
    var maiden = String.fromCharCode(nA + lon_div.round()) +
        String.fromCharCode(nA + lat_div.round());

    lat = lat_remainder;
    lon = lon_remainder / 2.0;

    var i = 1;
    // for every precision level we use one additional pair
    // first char in a pair presents lon, second lat
    while (i < precision) {
      i++;
      lon_div = lon.floor().toDouble();
      lon_remainder = (lon % 1).toDouble();
      lat_div = lat.floor().toDouble();
      lat_remainder = (lat % 1).toDouble();

      if ((i % 2 == 0)) {
        maiden += lon_div.round().toString() + lat_div.round().toString();
        lon = 24 * lon_remainder;
        lat = 24 * lat_remainder;
      } else {
        maiden += String.fromCharCode(nA + lon_div.round()) +
            String.fromCharCode(nA + lat_div.round());
        lon = 10 * lon_remainder;
        lat = 10 * lat_remainder;
      }

      // third pair is normally lowercase (e.g. 'JN75xu32SP')
      // we need to use toLowerCase
      if (maiden.length >= 6) {
        maiden = maiden.substring(0, 4) +
            maiden.substring(4, 6).toLowerCase() +
            maiden.substring(6);
      }
    }
    return maiden;
  }

  /// Convert maidenhead grid to the latitude and longitude
  /// Lat/Long with 6 decimal places
  static List<double> to_location(String grid) {
    // length is precision level * 2
    // if it is a standard most used precision 3 locator
    // length will be 6  (e.g. 'JN75xu')
    final lN = grid.length;
    final nA = 'A'.codeUnits.first;
    var lon = -180.0;
    var lat = -90.0;
    // we first need to convert to upper case
    final maiden = grid.toUpperCase();
    // extract first pair
    lon += (maiden[0].codeUnits.first - nA) * 20;
    lat += (maiden[1].codeUnits.first - nA) * 10;

    // second pair
    if (lN >= 4) {
      lon += int.parse(maiden[2]) * 2;
      lat += int.parse(maiden[3]) * 1;
    }

    // third pair (precision 3)
    if (lN >= 6) {
      lon += ((maiden[4].codeUnits.first - nA) * 5.0 / 60);
      lat += ((maiden[5].codeUnits.first - nA) * 2.5 / 60);
    }

    // fourth pair (precision 4)
    if (lN >= 8) {
      lon += int.parse(grid[6]) * 5.0 / 600;
      lat += int.parse(grid[7]) * 2.5 / 600;
    }

    // fifth pair (precision 5)
    if (lN >= 10) {
      lon += ((maiden[8].codeUnits.first - nA) / 2880);
      lat += ((maiden[9].codeUnits.first - nA) / 5760);
    }

    // truncate to 6 decimal places
    lat = round(lat, decimals: 6);
    lon = round(lon, decimals: 6);

    return List<double>.unmodifiable([lat, lon]);
  }

  /// Maidenhead grid to google maps url
  static String google_maps_maiden(String grid) {
    var latlon = to_location(grid);
    return 'https://www.google.com/maps/?api=1&map_action=map&center=' +
        latlon[0].toString() +
        ',' +
        latlon[1].toString();
  }

  double distance_to(Maidenhead p2) {
    return round(Haversine.distance(lat, lon, p2.lat, p2.lon), decimals: 2);
  }

  double bearing_to(Maidenhead p2) {
    return round(bearing(lat, lon, p2.lat, p2.lon), decimals: 2);
  }

  static double bearing(double lat, double lon, double lat2, double lon2) {
    final diffLon = degToRadian(lon2) - degToRadian(lon);
    final y = math.sin(diffLon);
    final x = math.cos(degToRadian(lat)) * math.tan(degToRadian(lat2)) -
        math.sin(degToRadian(lat)) * math.cos(diffLon);
    return radianToDeg(math.atan2(y, x));
  }
}


example/maidenhead_example.dart
import 'package:maidenhead/maidenhead.dart';

// nice online tool: https://k7fry.com/grid/

void main(List<String> arguments) {
  print('Expecting:  [45.844271, 15.947917]');
  print(Maidenhead.to_location('JN75xu32SP')); // 45.844271  15.947917

  print('Expecting: [45.845313 15.946181]');
  print(Maidenhead.to_location('JN75XU32NV')); // 45.845313 15.946181

  print('Expecting: JN75xu32SP');
  print(Maidenhead.to_maiden(45.844271, 15.947917, 5));

  // https://www.google.com/maps/?api=1&map_action=map&center=45.844271,15.947917
  print(Maidenhead.google_maps_maiden('JN75xu32SP'));

  print('Distance 177.69963804384477 meters');
  print(Haversine.distance(45.844271, 15.947917, 45.845313, 15.946181)
      .toString());

  // 177.69963804384477 meters
  print(Haversine.distance_maiden('JN75xu32SP', 'JN75XU32NV'));

  var pos1 = Maidenhead.latlon(45.844271, 15.947917, precision: 5);
  var pos2 = Maidenhead.latlon(45.844271, 15.947917, precision: 5);
  var distance = pos1.distance_to(pos2);
  var bearing = pos1.bearing_to(pos2);
  print('distance: $distance bearing: $bearing');
  print('pos1 to pos2 distance: ' +
      Haversine.distance_maiden(pos1.grid, pos2.grid).toString());

  print(
      'https://www.igismap.com/formula-to-find-bearing-or-heading-angle-between-two-points-latitude-longitude/');
  print(
      'K. City 39.099912, -94.581213 St. Louis 38.627089, -90.200203   96.51°');
  pos1 = Maidenhead.latlon(39.099912, -94.581213, precision: 5); // Kansas City
  pos2 = Maidenhead.latlon(38.627089, -90.200203, precision: 5); // St. Louis
  distance = pos1.distance_to(pos2);
  bearing = pos1.bearing_to(pos2);
  print('distance: $distance bearing K. City -> St. Louis: $bearing');
}
